Do not overwrite the dox field in Cargo.toml

pentamassiv opened the pull request:
I don't know how gir would know if a dependency has a dox feature without connecting to the internet or maintaining a list. This is why I opted for this simple solution:

Now gir just doesn't overwrite the dox field if it is present in the Cargo.toml. All other features will still be overwritten so new ones can be automatically added. This will fix https://github.com/gtk-rs/gir/issues/1381 and allows the merging of the PRs mentioned in the issue.

bilelmoussaoui commented on src/codegen/sys/cargo_toml.rs:
```diff
@@ -93,6 +93,20 @@ fn fill_empty(root: &mut Table, env: &Env, crate_name: &str) {
+        features.insert(
+            "dox".to_string(),
+            Value::Array(
+                env.config
```
I would actually drop this dox_feature_dependencies  configuration as nothing seems to use (gtk-rs-core/gstreamer-rs/gtk4-rs don't which are the largest users of gir) and gir is actually aware of the glib-based dependencies which all have a dox feature so you can generate the list automatically

pentamassiv commented:
Sure, I created a PR for fixing clippy lints:
https://github.com/gtk-rs/gir/pull/1383

pentamassiv commented:
I removed the commit that fixes the clippy lint and am now looking into implementing Bilals suggestion. I marked it as a draft for now

pentamassiv commented:
@bilelmoussaoui, is this what you wanted? 
If a `Gir.toml` contains the following:
```
external_libraries = [
    "ExternalLib1",
    "ExternalLib2",
    "ExternalLib3",
]
```
gir would now add this to the `Cargo.toml`, if it is not allready present: 
```
[features]
dox = [ "externalLib1/dox", "externalLib2/dox", "externalLib3/dox" ] 
```

Commit https://github.com/gtk-rs/gir/pull/1382/commits/f0a81972ce0f135939fec7da07fb0b4d374f8f57 removed the code that was relevant for dox_feature_dependencies.

bilelmoussaoui commented:
Would you mind generating gtk-rs-core with the changes of your PR and see if it works correctly?

sdroege commented on src/codegen/sys/cargo_toml.rs:
```diff
@@ -93,6 +93,15 @@ fn fill_empty(root: &mut Table, env: &Env, crate_name: &str) {
     }
+
+    {
+        let features = upsert_table(root, "features");
```
It doesn't seem to automatically add the `dox` feature with the corresponding dependencies if it's missing.

I tried removing the dox feature from pangocairo-sys and regenerated, but it wasn't added back.

sdroege replied:
Moving this to `fill_in` from `fill_empty` seems to do the right thing as @bilelmoussaoui suggested

```diff
diff --git a/gdk-pixbuf/sys/Cargo.toml b/gdk-pixbuf/sys/Cargo.toml
index 3508c36be21..bca362137e7 100644
--- a/gdk-pixbuf/sys/Cargo.toml
+++ b/gdk-pixbuf/sys/Cargo.toml
@@ -21,7 +21,7 @@ shell-words = "1.0.0"
 tempfile = "3"
 
 [features]
-dox = []
+dox = ["glib/dox", "gobject/dox", "gio/dox"]
 v2_40 = []
 v2_42 = ["v2_40"]
 
diff --git a/gio/sys/Cargo.toml b/gio/sys/Cargo.toml
index 29d349db48c..7dd6edcbe50 100644
--- a/gio/sys/Cargo.toml
+++ b/gio/sys/Cargo.toml
@@ -20,7 +20,7 @@ shell-words = "1.0.0"
 tempfile = "3"
 
 [features]
-dox = []
+dox = ["glib/dox", "gobject/dox"]
 v2_58 = []
 v2_60 = ["v2_58"]
 v2_62 = ["v2_60"]
diff --git a/glib/gobject-sys/Cargo.toml b/glib/gobject-sys/Cargo.toml
index ad64a68f693..3da746af848 100644
--- a/glib/gobject-sys/Cargo.toml
+++ b/glib/gobject-sys/Cargo.toml
@@ -13,7 +13,7 @@ shell-words = "1.0.0"
 tempfile = "3"
 
 [features]
-dox = []
+dox = ["glib/dox"]
 v2_58 = []
 v2_62 = ["v2_58"]
 v2_66 = ["v2_62"]
diff --git a/graphene/sys/Cargo.toml b/graphene/sys/Cargo.toml
index ba0d74cdea3..a3aac433317 100644
--- a/graphene/sys/Cargo.toml
+++ b/graphene/sys/Cargo.toml
@@ -14,7 +14,7 @@ shell-words = "1.0.0"
 tempfile = "3"
 
 [features]
-dox = []
+dox = ["glib/dox"]
 
 [lib]
 name = "graphene_sys"
diff --git a/pango/sys/Cargo.toml b/pango/sys/Cargo.toml
index 7cb0d098ec4..81df8950383 100644
--- a/pango/sys/Cargo.toml
+++ b/pango/sys/Cargo.toml
@@ -17,7 +17,7 @@ shell-words = "1.0.0"
 tempfile = "3"
 
 [features]
-dox = []
+dox = ["glib/dox", "gobject/dox"]
 v1_42 = []
 v1_44 = ["v1_42"]
 v1_46 = ["v1_44"]
diff --git a/pangocairo/sys/Cargo.toml b/pangocairo/sys/Cargo.toml
index 176d5ba174b..58979fc4cf7 100644
--- a/pangocairo/sys/Cargo.toml
+++ b/pangocairo/sys/Cargo.toml
@@ -21,7 +21,7 @@ shell-words = "1.0.0"
 tempfile = "3"
 
 [features]
-dox = []
+dox = ["glib/dox", "pango/dox", "cairo/dox"]
 
 [lib]
 name = "pango_cairo_sys"
```

sdroege commented:
```
Error: failed to get `gtk-sys` as a dependency of package `sys_build v0.0.1 (/home/runner/work/gir/gir/tests/sys/sys_build)`
Caused by:
  failed to load source for dependency `gtk-sys`

Caused by:
  Unable to update /home/runner/work/gir/gir/tests/sys/gtk-sys

Caused by:
  failed to parse manifest at `/home/runner/work/gir/gir/tests/sys/gtk-sys/Cargo.toml`

Caused by:
  feature `dox` includes `gdk_pixbuf/dox`, but `gdk_pixbuf` is not a dependency
```

Seems more like a bug in the test / gtk-rs though?

bilelmoussaoui commented:
> ```
> Error: failed to get `gtk-sys` as a dependency of package `sys_build v0.0.1 (/home/runner/work/gir/gir/tests/sys/sys_build)`
> Caused by:
>   failed to load source for dependency `gtk-sys`
> 
> Caused by:
>   Unable to update /home/runner/work/gir/gir/tests/sys/gtk-sys
> 
> Caused by:
>   failed to parse manifest at `/home/runner/work/gir/gir/tests/sys/gtk-sys/Cargo.toml`
> 
> Caused by:
>   feature `dox` includes `gdk_pixbuf/dox`, but `gdk_pixbuf` is not a dependency
> ```
> 
> Seems more like a bug in the test / gtk-rs though?

Because it has gdk_pixbuf renamed as gdk-pixbuf which doesn't exists. Not sure if renaming that here https://github.com/gtk-rs/gir/blob/master/tests/sys/gtk-sys/Cargo.toml#L17 would be enough though

pentamassiv commented:
I added a few lines to rename gdk_pixbuf/dox to gdk-pixbuf/dox. It is similar to the [renaming in the fill_empty method](https://github.com/gtk-rs/gir/blob/master/src/codegen/sys/cargo_toml.rs#L52).
I then cloned the [gtk-rs-core](https://github.com/gtk-rs/gtk-rs-core) repo and replaced the gir folder in it with the gir folder with the patches of this PR. Then I ran `python3 generator.py`. The only changes were in the versions.txt files and Cargo.toml files. The changes look good to me and `cargo build --all-features` succeeds without any errors.
Here is the full diff:
```
diff --git a/gdk-pixbuf/src/auto/versions.txt b/gdk-pixbuf/src/auto/versions.txt
index 7041be446..b11e29400 100644
--- a/gdk-pixbuf/src/auto/versions.txt
+++ b/gdk-pixbuf/src/auto/versions.txt
@@ -1,2 +1,2 @@
-Generated by gir (https://github.com/gtk-rs/gir @ b3147f2b6043)
+Generated by gir (https://github.com/gtk-rs/gir @ 10e894a8f0f9)
 from gir-files (https://github.com/gtk-rs/gir-files @ 7fa401e3ee5d)
diff --git a/gdk-pixbuf/sys/Cargo.toml b/gdk-pixbuf/sys/Cargo.toml
index 3508c36be..bca362137 100644
--- a/gdk-pixbuf/sys/Cargo.toml
+++ b/gdk-pixbuf/sys/Cargo.toml
@@ -21,7 +21,7 @@ shell-words = "1.0.0"
 tempfile = "3"
 
 [features]
-dox = []
+dox = ["glib/dox", "gobject/dox", "gio/dox"]
 v2_40 = []
 v2_42 = ["v2_40"]
 
diff --git a/gdk-pixbuf/sys/versions.txt b/gdk-pixbuf/sys/versions.txt
index 7041be446..b11e29400 100644
--- a/gdk-pixbuf/sys/versions.txt
+++ b/gdk-pixbuf/sys/versions.txt
@@ -1,2 +1,2 @@
-Generated by gir (https://github.com/gtk-rs/gir @ b3147f2b6043)
+Generated by gir (https://github.com/gtk-rs/gir @ 10e894a8f0f9)
 from gir-files (https://github.com/gtk-rs/gir-files @ 7fa401e3ee5d)
diff --git a/gio/src/auto/versions.txt b/gio/src/auto/versions.txt
index 7041be446..b11e29400 100644
--- a/gio/src/auto/versions.txt
+++ b/gio/src/auto/versions.txt
@@ -1,2 +1,2 @@
-Generated by gir (https://github.com/gtk-rs/gir @ b3147f2b6043)
+Generated by gir (https://github.com/gtk-rs/gir @ 10e894a8f0f9)
 from gir-files (https://github.com/gtk-rs/gir-files @ 7fa401e3ee5d)
diff --git a/gio/sys/Cargo.toml b/gio/sys/Cargo.toml
index 29d349db4..7dd6edcbe 100644
--- a/gio/sys/Cargo.toml
+++ b/gio/sys/Cargo.toml
@@ -20,7 +20,7 @@ shell-words = "1.0.0"
 tempfile = "3"
 
 [features]
-dox = []
+dox = ["glib/dox", "gobject/dox"]
 v2_58 = []
 v2_60 = ["v2_58"]
 v2_62 = ["v2_60"]
diff --git a/gio/sys/versions.txt b/gio/sys/versions.txt
index 7041be446..b11e29400 100644
--- a/gio/sys/versions.txt
+++ b/gio/sys/versions.txt
@@ -1,2 +1,2 @@
-Generated by gir (https://github.com/gtk-rs/gir @ b3147f2b6043)
+Generated by gir (https://github.com/gtk-rs/gir @ 10e894a8f0f9)
 from gir-files (https://github.com/gtk-rs/gir-files @ 7fa401e3ee5d)
diff --git a/gir b/gir
index b3147f2b6..10e894a8f 160000
--- a/gir
+++ b/gir
@@ -1 +1 @@
-Subproject commit b3147f2b6043ee273283dc45832da8757ab654d6
+Subproject commit 10e894a8f0f9001f59f6376567c34424cc8b33d1
diff --git a/glib/gobject-sys/Cargo.toml b/glib/gobject-sys/Cargo.toml
index ad64a68f6..3da746af8 100644
--- a/glib/gobject-sys/Cargo.toml
+++ b/glib/gobject-sys/Cargo.toml
@@ -13,7 +13,7 @@ shell-words = "1.0.0"
 tempfile = "3"
 
 [features]
-dox = []
+dox = ["glib/dox"]
 v2_58 = []
 v2_62 = ["v2_58"]
 v2_66 = ["v2_62"]
diff --git a/glib/gobject-sys/versions.txt b/glib/gobject-sys/versions.txt
index 7041be446..b11e29400 100644
--- a/glib/gobject-sys/versions.txt
+++ b/glib/gobject-sys/versions.txt
@@ -1,2 +1,2 @@
-Generated by gir (https://github.com/gtk-rs/gir @ b3147f2b6043)
+Generated by gir (https://github.com/gtk-rs/gir @ 10e894a8f0f9)
 from gir-files (https://github.com/gtk-rs/gir-files @ 7fa401e3ee5d)
diff --git a/glib/src/auto/versions.txt b/glib/src/auto/versions.txt
index 7041be446..b11e29400 100644
--- a/glib/src/auto/versions.txt
+++ b/glib/src/auto/versions.txt
@@ -1,2 +1,2 @@
-Generated by gir (https://github.com/gtk-rs/gir @ b3147f2b6043)
+Generated by gir (https://github.com/gtk-rs/gir @ 10e894a8f0f9)
 from gir-files (https://github.com/gtk-rs/gir-files @ 7fa401e3ee5d)
diff --git a/glib/sys/versions.txt b/glib/sys/versions.txt
index 7041be446..b11e29400 100644
--- a/glib/sys/versions.txt
+++ b/glib/sys/versions.txt
@@ -1,2 +1,2 @@
-Generated by gir (https://github.com/gtk-rs/gir @ b3147f2b6043)
+Generated by gir (https://github.com/gtk-rs/gir @ 10e894a8f0f9)
 from gir-files (https://github.com/gtk-rs/gir-files @ 7fa401e3ee5d)
diff --git a/graphene/src/auto/versions.txt b/graphene/src/auto/versions.txt
index 7041be446..b11e29400 100644
--- a/graphene/src/auto/versions.txt
+++ b/graphene/src/auto/versions.txt
@@ -1,2 +1,2 @@
-Generated by gir (https://github.com/gtk-rs/gir @ b3147f2b6043)
+Generated by gir (https://github.com/gtk-rs/gir @ 10e894a8f0f9)
 from gir-files (https://github.com/gtk-rs/gir-files @ 7fa401e3ee5d)
diff --git a/graphene/sys/Cargo.toml b/graphene/sys/Cargo.toml
index ba0d74cde..a3aac4333 100644
--- a/graphene/sys/Cargo.toml
+++ b/graphene/sys/Cargo.toml
@@ -14,7 +14,7 @@ shell-words = "1.0.0"
 tempfile = "3"
 
 [features]
-dox = []
+dox = ["glib/dox"]
 
 [lib]
 name = "graphene_sys"
diff --git a/graphene/sys/versions.txt b/graphene/sys/versions.txt
index 7041be446..b11e29400 100644
--- a/graphene/sys/versions.txt
+++ b/graphene/sys/versions.txt
@@ -1,2 +1,2 @@
-Generated by gir (https://github.com/gtk-rs/gir @ b3147f2b6043)
+Generated by gir (https://github.com/gtk-rs/gir @ 10e894a8f0f9)
 from gir-files (https://github.com/gtk-rs/gir-files @ 7fa401e3ee5d)
diff --git a/pango/src/auto/versions.txt b/pango/src/auto/versions.txt
index 7041be446..b11e29400 100644
--- a/pango/src/auto/versions.txt
+++ b/pango/src/auto/versions.txt
@@ -1,2 +1,2 @@
-Generated by gir (https://github.com/gtk-rs/gir @ b3147f2b6043)
+Generated by gir (https://github.com/gtk-rs/gir @ 10e894a8f0f9)
 from gir-files (https://github.com/gtk-rs/gir-files @ 7fa401e3ee5d)
diff --git a/pango/sys/Cargo.toml b/pango/sys/Cargo.toml
index 7cb0d098e..81df89503 100644
--- a/pango/sys/Cargo.toml
+++ b/pango/sys/Cargo.toml
@@ -17,7 +17,7 @@ shell-words = "1.0.0"
 tempfile = "3"
 
 [features]
-dox = []
+dox = ["glib/dox", "gobject/dox"]
 v1_42 = []
 v1_44 = ["v1_42"]
 v1_46 = ["v1_44"]
diff --git a/pango/sys/versions.txt b/pango/sys/versions.txt
index 7041be446..b11e29400 100644
--- a/pango/sys/versions.txt
+++ b/pango/sys/versions.txt
@@ -1,2 +1,2 @@
-Generated by gir (https://github.com/gtk-rs/gir @ b3147f2b6043)
+Generated by gir (https://github.com/gtk-rs/gir @ 10e894a8f0f9)
 from gir-files (https://github.com/gtk-rs/gir-files @ 7fa401e3ee5d)
diff --git a/pangocairo/src/auto/versions.txt b/pangocairo/src/auto/versions.txt
index 7041be446..b11e29400 100644
--- a/pangocairo/src/auto/versions.txt
+++ b/pangocairo/src/auto/versions.txt
@@ -1,2 +1,2 @@
-Generated by gir (https://github.com/gtk-rs/gir @ b3147f2b6043)
+Generated by gir (https://github.com/gtk-rs/gir @ 10e894a8f0f9)
 from gir-files (https://github.com/gtk-rs/gir-files @ 7fa401e3ee5d)
diff --git a/pangocairo/sys/Cargo.toml b/pangocairo/sys/Cargo.toml
index 176d5ba17..58979fc4c 100644
--- a/pangocairo/sys/Cargo.toml
+++ b/pangocairo/sys/Cargo.toml
@@ -21,7 +21,7 @@ shell-words = "1.0.0"
 tempfile = "3"
 
 [features]
-dox = []
+dox = ["glib/dox", "pango/dox", "cairo/dox"]
 
 [lib]
 name = "pango_cairo_sys"
diff --git a/pangocairo/sys/versions.txt b/pangocairo/sys/versions.txt
index 7041be446..b11e29400 100644
--- a/pangocairo/sys/versions.txt
+++ b/pangocairo/sys/versions.txt
@@ -1,2 +1,2 @@
-Generated by gir (https://github.com/gtk-rs/gir @ b3147f2b6043)
+Generated by gir (https://github.com/gtk-rs/gir @ 10e894a8f0f9)
 from gir-files (https://github.com/gtk-rs/gir-files @ 7fa401e3ee5d)

```

pentamassiv commented:
Should the minor version be increased with this merged?
I think this would be a breaking change since using the generated code in combination with the `--all-features` flag would lead to failures since the libraries would no longer be linked.

sdroege commented on src/codegen/sys/cargo_toml.rs:
```diff
@@ -186,7 +186,12 @@ fn fill_in(root: &mut Table, env: &Env) {
         let mut dox_features = Vec::new();
         for ext_lib in &env.config.external_libraries {
-            dox_features.push(Value::String(format!("{}/dox", ext_lib.crate_name)));
+            let crate_name = if ext_lib.crate_name == "gdk_pixbuf" {
```
Should probably replicate the whole block here then to also cover cairo? And move it to a separate function :)

```rust
        let ext_package = if ext_lib.crate_name == "cairo" {
        let ext_package = if ext_lib.crate_name == "cairo" {
            format!("{}-sys-rs", ext_lib.crate_name)
            format!("{}-sys-rs", ext_lib.crate_name)
        } else if ext_lib.crate_name == "gdk_pixbuf" {
        } else if ext_lib.crate_name == "gdk_pixbuf" {
            "gdk-pixbuf-sys".into()
            "gdk-pixbuf-sys".into()
        } else {
        } else {
            format!("{}-sys", ext_lib.crate_name)
            format!("{}-sys", ext_lib.crate_name)
        };
```

pentamassiv replied:
I don't think other renaming is necessary. 

1. The cairo dependency is renamed in the sys crates to just "cairo" (e.g in [pangocairo](https://github.com/gtk-rs/gtk-rs-core/blob/master/pangocairo/sys/Cargo.toml#L7)).
2. The gdk_pixbuf renaming is being done
3. the rest is just to add the -sys suffix but in the -sys crates, the dependencies are renamed and the suffix is not needed ([example](https://github.com/gtk-rs/gtk-rs-core/blob/master/pangocairo/sys/Cargo.toml#L11))

Since the gdk_pixbuf renaming is the only one that is needed for the dox feature, I decided it would be easiest to copy the code and not move it to a separate function. I can change that if you'd like me too though

sdroege replied:
Maybe gdkpixbuf should also be renamed like cairo then so this special case here is not needed? Or is that not easily possible?

pentamassiv replied:
As far as I can tell all other crates related to gtk use the - instead of _ in their names. I don't know why it was decided to do the renaming, but my guess is it was done to have a consistent naming scheme.

pentamassiv replied:
Since it is generated and a dependency in the sys crates, I don't think the name is very important.

pentamassiv replied:
Just to clarify, which name were you proposing instead of `gdk_pixbuf`?

bilelmoussaoui replied:
If we can avoid special casing gdk-pixbuf here that would be for the better for sure.

sdroege commented:
> Should the minor version be increased with this merged?
> [...]

The crate versions are already `0.16.0` (etc), so that's already managed. The last release was 0.15 :)

bilelmoussaoui commented:
The changes looks good to me personally, would you mind squashing the commits into one please?

sdroege commented:
Let's get it in then? I assume you all confirmed that it doesn't break any of our crates? :)

pentamassiv commented:
Tomorrow I'll test it with gtk3, gtk4 and libadwaita. So far I tested it with gtk-rs-core.

Also I noticed that there are a bunch of files in the gir/tests folder which contain
```
[features]
dox = []
```
I guess I have to manually fix those and make them contain the same line as in their respective crate to ensure the tests use the same Cargo.toml.

sdroege commented:
> Tomorrow I'll test it with gtk3, gtk4 and libadwaita. So far I tested it with gtk-rs-core.

Thanks!

> Also I noticed that there are a bunch of files in the gir/tests folder which contain

Maybe those should just be updated with the real ones

bilelmoussaoui commented:
This one is no longer needed

sdroege commented:
It's superseded by https://github.com/gtk-rs/gir/pull/1388 (which is based on this one)
